### PR TITLE
[Lang] MatrixField refactor 5/n: Lower access of matrix field element into CHI IR

### PR DESCRIPTION
**Pull request by strongoier**
Issue: #5959

### Brief Summary

This PR lowers access of matrix field element into CHI IR. To verify its functionality, this PR also makes the following path of #5959 running end-to-end (currently only enabled when `real_matrix=True`):

MatrixField_Part5

Key steps:
- Python: when running into subscript of `MatrixField`, create `IndexExpression` in C++ instead of building `_MatrixFieldElement` in Python
- Frontend IR: allow `IndexExpression` to take `MatrixFieldExpression`
- CHI IR: add `MatrixOfGlobalPtrStmt`; add `lower_matrix_ptr` pass, which demotes `PtrOffsetStmt(MatrixOfGlobalPtrStmt)` into `GlobalPtrStmt` in the case of constant index - this is incomplete and only aims at supporting the above path

### CODE REVIEW

File: taichi/ir/expr.cpp [hunk 1]
```diff
@@ -31,8 +31,8 @@ Expr Expr::operator[](const ExprGroup &indices) const {
     // Allow indexing IndexExpression with ret_type = TensorType
     TI_ASSERT(is_tensor(expr->ret_type));
   } else {
-    TI_ASSERT(is<FieldExpression>() || is<ExternalTensorExpression>() ||
-              is<IdExpression>());
+    TI_ASSERT(is<FieldExpression>() || is<MatrixFieldExpression>() ||
```
Comment (AD1024) [inline taichi/ir/expr.cpp:34]: Would this check disable usage such as `(a @ b)[index]`?

File: taichi/ir/frontend_ir.cpp [hunk 2]
```diff
@@ -401,7 +399,30 @@ Stmt *make_field_access(Expression::FlattenContext *ctx,
...
+  SNode *snode = var.cast<FieldExpression>()->snode;
+  return ctx->push_back(std::make_unique<GlobalPtrStmt>(
+      snode, make_index_stmts(ctx, indices, snode->index_offsets)));
+}
+
+Stmt *make_matrix_field_access(Expression::FlattenContext *ctx,
+                               Expr var,
+                               ExprGroup indices,
+                               DataType ret_type) {
+  auto matrix_field = var.cast<MatrixFieldExpression>();
```
Comment (jim19930609) [inline taichi/ir/frontend_ir.cpp:417]: assert `matrix_field != nullptr` after dynamic pointer cast? Or take `MatrixFieldExpression var` explicitly?

File: taichi/transforms/lower_matrix_ptr.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+TLANG_NAMESPACE_BEGIN
+
+class LowerMatrixPtr : public BasicStmtVisitor {
+ public:
+  using BasicStmtVisitor::visit;
+  DelayedIRModifier modifier;
+
+  void visit(PtrOffsetStmt *stmt) override {
+    if (stmt->origin->is<MatrixOfGlobalPtrStmt>()) {
+      TI_ASSERT(stmt->offset->is<ConstStmt>());
```
Comment (jim19930609) [inline taichi/transforms/lower_matrix_ptr.cpp:16]: What about the situation where a MatrixField is indexed with a LoopIndex? for example:
```
x = ti.Matrix.field(m=2, shape=(5))
@ti.kernel
def test():
  for i in x:
    for j in range(2):
       x[i][j] = 100
```
In this case, x[i] is MatrixOfGlobalPtrStmt, and x[i][j] is a PtrOffsetStmt with offset being `LoopIndexStmt`?

File: tests/python/test_matrix.py [hunk 4]
```diff
@@ -248,8 +248,7 @@ def func4(k: ti.i32):
     func4(10)
 
 
-@test_utils.test(arch=ti.cpu)
-def test_matrix_constant_index():
+def _test_matrix_constant_index():
```
Comment (jim19930609) [inline tests/python/test_matrix.py:251]: Maybe add some tests with elementwise operations involved in a following PR~

File: taichi/transforms/lower_matrix_ptr.cpp (hunk 3 above)
Comment (strongoier) [reply taichi/transforms/lower_matrix_ptr.cpp:16]: This case requires dynamic index, which will be supported in a future PR :-)

File: taichi/ir/expr.cpp (hunk 1 above)
Comment (strongoier) [reply taichi/ir/expr.cpp]: Nice catch. Fixed.

File: taichi/ir/frontend_ir.cpp (hunk 2 above)
Comment (strongoier) [reply taichi/ir/frontend_ir.cpp]: Sure. I've chosen the latter.

**Review by jim19930609 — APPROVED** [review]
LGTM!

**Pull request merged**